compiler/src/dotty/tools/dotc/core/TypeOps.scala

```diff
@@ -712,7 +712,16 @@ object TypeOps:
 
     val childTp = if (child.isTerm) child.termRef else child.typeRef
 
-    inContext(ctx.fresh.setExploreTyperState().setFreshGADTBounds) {
+    val ctx1 = ctx.fresh.setExploreTyperState().setFreshGADTBounds
+    val ctx2 = parent match
+      case _: RefinedType =>
+        ctx1
+        // patmat/t9657
+        // When running Bicycle.type <:< Vehicle { A = P }
+        // TypeComparer is happy to infer GADT bounds P >: Pedal.type <: Petrol.type & Pedal.type
+        // Despite the fact that Bicycle is an object, and thus final, so its type A can only be Pedal.type.
+      case _              => ctx1.addMode(Mode.GadtConstraintInference)
```

Comment:
That PR does not seem to fix this. I found that the direct cause is: at some point, the inferred GADT bounds is impossible (Petal.type <:< Petrol.type is absurd), but we do not rollback this bad constraint and continue with it, resulting in false comparison result. What happens specifically:

- We call `narrowGADTBounds` [here](https://github.com/lampepfl/dotty/blob/main/compiler/src/dotty/tools/dotc/core/TypeComparer.scala#L536), this function returns `false` after recording the bound `P >: Pedal.type` since the resulting bounds `P: >: Pedal.type <: Petrol.type & Pedal.type` is impossible.

- This makes `compareGADT` fail, but we do not rollback this bad constraint, and continue with other tries: `tryLiftToThis2` and `fourthTry`.

- `fourthTry` succeeds with this bad constraint.

I think the root cause might lie in the difference of the assumption of GADT reasoning and patmat checking.

- When comparing a subtype relationship `S <:< T` when doing GADT reasoning, we *assume* that this relationship holds, and look for necessary constraints resulted from this relationship. Therefore, when `compareGADT` gives us an impossible constraint necessarily following from the relationship `S <:< T`, this implies that the relationship `S <:< T` itself is absurd, so it would be sound to continue other attempts with this bad constraint. Rolling back is not necessary.

- However, when comparing `S <:< T` during patmat checking, we essentially checks the satisfiability of this relationship. We can not assume it true. In this case, not rolling back the impossible GADT constriant gives us incorrect comparison result.

It seems that rolling back GADT constraints when `narrowGADTBounds` returns `false` could solve this issue. But I am not 100% sure whether this will affect GADT reasoning.